android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java

```diff
@@ -312,18 +311,21 @@ protected void onResume() {
             // URL contains KMP to download in background.
             boolean isCustom = KMKeyboardDownloaderActivity.isCustom(url);
 
-            int index = url.lastIndexOf("/") + 1;
-            String filename = "unknown:";
-            if (index >= 0 && index <= url.length()) {
-              filename = url.substring(index);
+            String filename = data.getQueryParameter("filename");
+            if (filename == null) {
+              int index = url.lastIndexOf("/") + 1;
+              if (index >= 0 && index <= url.length()) {
+                filename = url.substring(index);
```

Comment:
Is there a possibility of url encoding tripping us up (e.g. `%20` for space)?

Reply:
It looks like URI handles either.

I tried these download links:
```
keyman://?filename=khmer_angkor.kmp&url=https%3A%2F%2Fkeyman.com%2Fkeyboard%2Fdownload%3Fid%3Dkhmer_angkor%26platform%3Dandroid

keyman://?filename=khmer_angkor.kmp&url=https://keyman.com/keyboard/download?id=khmer_angkor%26%20platform=android
```
And both downloaded khmer_angkor.kmp fine and installed it.